File: ts/src/any_llm.ts
import type {
  ChatCompletion,
  ChatCompletionChunk,
  ChatCompletionMessageParam,
} from "openai/resources/chat/completions";
import type {
  CompletionOptions,
  CompletionParams,
  LLMProvider,
} from "./types.js";

/**
 * Abstract base class for all LLM providers.
 * All providers must inherit from this class and implement the abstract methods.
 */
export abstract class AnyLLM {
  /**
   * Name of the provider (e.g., "openai", "anthropic").
   */
  abstract readonly providerName: string;

  /**
   * Create a provider-specific AnyLLM instance.
   * Providers are loaded dynamically with lazy imports.
   *
   * @param provider - The provider name or LLMProvider enum value
   * @param apiKey - Optional API key (falls back to environment variable)
   * @param apiBase - Optional API base URL
   * @param options - Additional provider-specific options
   * @returns A provider-specific AnyLLM instance
   */
  static async create(
    provider: string | LLMProvider,
    apiKey?: string,
    apiBase?: string,
    options?: Record<string, unknown>,
  ): Promise<AnyLLM> {
    const providerName = provider.toLowerCase();

    switch (providerName) {
      case "openai": {
        const { OpenAIProvider } = await import("./providers/openai.js");
        const instance = new OpenAIProvider();
        instance.initClient(apiKey, apiBase, options);
        return instance;
      }
      default:
        throw new Error(`Unknown provider: ${provider}`);
    }
  }

  /**
   * Initialize the provider's HTTP/SDK client.
   * When apiKey is undefined, should check for standard environment variables.
   *
   * @param apiKey - Optional API key
   * @param apiBase - Optional API base URL
   * @param options - Additional provider-specific options
   */
  abstract initClient(
    apiKey?: string,
    apiBase?: string,
    options?: Record<string, unknown>,
  ): void;

  /**
   * Perform a chat completion request.
   *
   * @param model - The model ID to use
   * @param messages - The conversation messages
   * @param options - Optional completion parameters
   * @returns The chat completion response
   */
  abstract completion(
    model: string,
    messages: ChatCompletionMessageParam[],
    options?: CompletionOptions,
  ): Promise<ChatCompletion>;

  /**
   * Perform a streaming chat completion request.
   *
   * @param model - The model ID to use
   * @param messages - The conversation messages
   * @param options - Optional completion parameters
   * @returns An async iterable of chat completion chunks
   */
  abstract completionStream(
    model: string,
    messages: ChatCompletionMessageParam[],
    options?: CompletionOptions,
  ): AsyncIterable<ChatCompletionChunk>;

  /**
   * Convert user-friendly CompletionOptions to provider-specific parameters.
   *
   * @param params - The internal completion parameters
   * @param extraOptions - Additional provider-specific options
   * @returns Provider-specific request parameters
   */
  protected static convertCompletionParams(
    params: CompletionParams,
    _extraOptions?: Record<string, unknown>,
  ): Record<string, unknown> {
    return { ...params };
  }

  /**
   * Convert provider-specific response to standard ChatCompletion.
   *
   * @param response - The provider-specific response
   * @returns Standard ChatCompletion response
   */
  protected static convertCompletionResponse(
    response: unknown,
  ): ChatCompletion {
    return response as ChatCompletion;
  }

  /**
   * Convert provider-specific chunk to standard ChatCompletionChunk.
   *
   * @param chunk - The provider-specific chunk
   * @returns Standard ChatCompletionChunk
   */
  protected static convertCompletionChunkResponse(
    chunk: unknown,
  ): ChatCompletionChunk {
    return chunk as ChatCompletionChunk;
  }

  /**
   * Build internal CompletionParams from user options.
   *
   * @param model - The model ID
   * @param messages - The conversation messages
   * @param options - User-friendly completion options
   * @param stream - Whether to enable streaming
   * @returns Internal completion parameters
   */
  protected buildCompletionParams(
    model: string,
    messages: ChatCompletionMessageParam[],
    options?: CompletionOptions,
    stream?: boolean,
  ): CompletionParams {
    const params: CompletionParams = {
      model,
      messages,
    };

    if (stream !== undefined) {
      params.stream = stream;
    }

    if (options) {
      if (options.temperature !== undefined) {
        params.temperature = options.temperature;
      }
      if (options.topP !== undefined) {
        params.top_p = options.topP;
      }
      if (options.maxTokens !== undefined) {
        params.max_tokens = options.maxTokens;
      }
      if (options.maxCompletionTokens !== undefined) {
        params.max_completion_tokens = options.maxCompletionTokens;
      }
      if (options.n !== undefined) {
        params.n = options.n;
      }
      if (options.stop !== undefined) {
        params.stop = options.stop;
      }
      if (options.presencePenalty !== undefined) {
        params.presence_penalty = options.presencePenalty;
      }
      if (options.frequencyPenalty !== undefined) {
        params.frequency_penalty = options.frequencyPenalty;
      }
      if (options.logitBias !== undefined) {
        params.logit_bias = options.logitBias;
      }
      if (options.user !== undefined) {
        params.user = options.user;
      }
      if (options.tools !== undefined) {
        params.tools = options.tools;
      }
      if (options.toolChoice !== undefined) {
        params.tool_choice = options.toolChoice;
      }
      if (options.parallelToolCalls !== undefined) {
        params.parallel_tool_calls = options.parallelToolCalls;
      }
      if (options.responseFormat !== undefined) {
        params.response_format = options.responseFormat;
      }
      if (options.logprobs !== undefined) {
        params.logprobs = options.logprobs;
      }
      if (options.topLogprobs !== undefined) {
        params.top_logprobs = options.topLogprobs;
      }
      if (options.streamOptions !== undefined) {
        params.stream_options = {
          include_usage: options.streamOptions.includeUsage,
        };
      }
    }

    return params;
  }
}

File: py/src/any_llm/any_llm.py
from abc import ABC, abstractmethod
from collections.abc import AsyncIterator, Iterable, Iterator
from typing import Any

from any_llm.providers import Providers, get_provider_class
from any_llm.types import (
    ChatCompletion,
    ChatCompletionChunk,
    ChatCompletionMessageParam,
    ChatCompletionToolParam,
    CompletionParams,
)
from any_llm.utils.aio import async_iter_to_sync_iter, run_async_in_sync


class AnyLLM(ABC):
    """Abstract base class for LLM providers.

    All providers must inherit from this class and implement the abstract methods.
    """

    @classmethod
    async def acreate(
        cls,
        provider: str | Providers,
        api_key: str | None = None,
        api_base: str | None = None,
        **kwargs: Any,
    ) -> "AnyLLM":
        """Create a provider-specific LLM client asynchronously.

        Args:
            provider: The provider name or Providers enum value.
            api_key: API key for the provider. If None, will check environment variables.
            api_base: Base URL for the API. If None, uses provider default.
            **kwargs: Additional provider-specific arguments.

        Returns:
            An initialized LLM client for the specified provider.

        """
        provider_class = get_provider_class(provider)
        instance: AnyLLM = provider_class()
        instance._init_client(api_key=api_key, api_base=api_base, **kwargs)
        return instance

    @classmethod
    def create(
        cls,
        provider: str | Providers,
        api_key: str | None = None,
        api_base: str | None = None,
        **kwargs: Any,
    ) -> "AnyLLM":
        """Create a provider-specific LLM client synchronously.

        Args:
            provider: The provider name or Providers enum value.
            api_key: API key for the provider. If None, will check environment variables.
            api_base: Base URL for the API. If None, uses provider default.
            **kwargs: Additional provider-specific arguments.

        Returns:
            An initialized LLM client for the specified provider.

        """
        return run_async_in_sync(cls.acreate(provider, api_key, api_base, **kwargs))

    @abstractmethod
    def _init_client(
        self,
        api_key: str | None = None,
        api_base: str | None = None,
        **kwargs: Any,
    ) -> None:
        """Initialize the HTTP/Provider SDK client.

        Args:
            api_key: API key for the provider. If None, should check environment variables.
            api_base: Base URL for the API. If None, uses provider default.
            **kwargs: Additional provider-specific arguments.

        """

    @abstractmethod
    async def acompletion(
        self,
        model: str,
        messages: Iterable[ChatCompletionMessageParam],
        *,
        frequency_penalty: float | None = None,
        logit_bias: dict[str, int] | None = None,
        logprobs: bool | None = None,
        max_completion_tokens: int | None = None,
        max_tokens: int | None = None,
        n: int | None = None,
        parallel_tool_calls: bool | None = None,
        presence_penalty: float | None = None,
        response_format: dict[str, Any] | None = None,
        seed: int | None = None,
        stop: str | list[str] | None = None,
        stream_options: dict[str, Any] | None = None,
        temperature: float | None = None,
        tool_choice: str | dict[str, Any] | None = None,
        tools: Iterable[ChatCompletionToolParam] | None = None,
        top_logprobs: int | None = None,
        top_p: float | None = None,
        user: str | None = None,
        **kwargs: Any,
    ) -> ChatCompletion:
        """Perform a chat completion request asynchronously.

        Args:
            model: Model ID to use for the completion.
            messages: List of messages comprising the conversation.
            frequency_penalty: Penalize repeated tokens (-2.0 to 2.0).
            logit_bias: Map of token IDs to bias values.
            logprobs: Whether to return log probabilities.
            max_completion_tokens: Maximum tokens to generate.
            max_tokens: Legacy parameter for maximum tokens.
            n: Number of completions to generate.
            parallel_tool_calls: Enable parallel function calling.
            presence_penalty: Penalize new topics (-2.0 to 2.0).
            response_format: Format specification for structured outputs.
            seed: Seed for deterministic sampling.
            stop: Stop sequences.
            stream_options: Options for streaming.
            temperature: Sampling temperature (0 to 2).
            tool_choice: Controls tool calling behavior.
            tools: List of tools the model may call.
            top_logprobs: Number of top tokens to return.
            top_p: Nucleus sampling threshold.
            user: Unique user identifier.
            **kwargs: Additional provider-specific arguments.

        Returns:
            ChatCompletion response object.

        """

    def completion(
        self,
        model: str,
        messages: Iterable[ChatCompletionMessageParam],
        *,
        frequency_penalty: float | None = None,
        logit_bias: dict[str, int] | None = None,
        logprobs: bool | None = None,
        max_completion_tokens: int | None = None,
        max_tokens: int | None = None,
        n: int | None = None,
        parallel_tool_calls: bool | None = None,
        presence_penalty: float | None = None,
        response_format: dict[str, Any] | None = None,
        seed: int | None = None,
        stop: str | list[str] | None = None,
        stream_options: dict[str, Any] | None = None,
        temperature: float | None = None,
        tool_choice: str | dict[str, Any] | None = None,
        tools: Iterable[ChatCompletionToolParam] | None = None,
        top_logprobs: int | None = None,
        top_p: float | None = None,
        user: str | None = None,
        **kwargs: Any,
    ) -> ChatCompletion:
        """Perform a chat completion request synchronously.

        Args:
            model: Model ID to use for the completion.
            messages: List of messages comprising the conversation.
            frequency_penalty: Penalize repeated tokens (-2.0 to 2.0).
            logit_bias: Map of token IDs to bias values.
            logprobs: Whether to return log probabilities.
            max_completion_tokens: Maximum tokens to generate.
            max_tokens: Legacy parameter for maximum tokens.
            n: Number of completions to generate.
            parallel_tool_calls: Enable parallel function calling.
            presence_penalty: Penalize new topics (-2.0 to 2.0).
            response_format: Format specification for structured outputs.
            seed: Seed for deterministic sampling.
            stop: Stop sequences.
            stream_options: Options for streaming.
            temperature: Sampling temperature (0 to 2).
            tool_choice: Controls tool calling behavior.
            tools: List of tools the model may call.
            top_logprobs: Number of top tokens to return.
            top_p: Nucleus sampling threshold.
            user: Unique user identifier.
            **kwargs: Additional provider-specific arguments.

        Returns:
            ChatCompletion response object.

        """
        return run_async_in_sync(
            self.acompletion(
                model=model,
                messages=messages,
                frequency_penalty=frequency_penalty,
                logit_bias=logit_bias,
                logprobs=logprobs,
                max_completion_tokens=max_completion_tokens,
                max_tokens=max_tokens,
                n=n,
                parallel_tool_calls=parallel_tool_calls,
                presence_penalty=presence_penalty,
                response_format=response_format,
                seed=seed,
                stop=stop,
                stream_options=stream_options,
                temperature=temperature,
                tool_choice=tool_choice,
                tools=tools,
                top_logprobs=top_logprobs,
                top_p=top_p,
                user=user,
                **kwargs,
            )
        )

    @abstractmethod
    async def acompletion_stream(
        self,
        model: str,
        messages: Iterable[ChatCompletionMessageParam],
        *,
        frequency_penalty: float | None = None,
        logit_bias: dict[str, int] | None = None,
        logprobs: bool | None = None,
        max_completion_tokens: int | None = None,
        max_tokens: int | None = None,
        n: int | None = None,
        parallel_tool_calls: bool | None = None,
        presence_penalty: float | None = None,
        response_format: dict[str, Any] | None = None,
        seed: int | None = None,
        stop: str | list[str] | None = None,
        stream_options: dict[str, Any] | None = None,
        temperature: float | None = None,
        tool_choice: str | dict[str, Any] | None = None,
        tools: Iterable[ChatCompletionToolParam] | None = None,
        top_logprobs: int | None = None,
        top_p: float | None = None,
        user: str | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[ChatCompletionChunk]:
        """Perform a streaming chat completion request asynchronously.

        Args:
            model: Model ID to use for the completion.
            messages: List of messages comprising the conversation.
            frequency_penalty: Penalize repeated tokens (-2.0 to 2.0).
            logit_bias: Map of token IDs to bias values.
            logprobs: Whether to return log probabilities.
            max_completion_tokens: Maximum tokens to generate.
            max_tokens: Legacy parameter for maximum tokens.
            n: Number of completions to generate.
            parallel_tool_calls: Enable parallel function calling.
            presence_penalty: Penalize new topics (-2.0 to 2.0).
            response_format: Format specification for structured outputs.
            seed: Seed for deterministic sampling.
            stop: Stop sequences.
            stream_options: Options for streaming.
            temperature: Sampling temperature (0 to 2).
            tool_choice: Controls tool calling behavior.
            tools: List of tools the model may call.
            top_logprobs: Number of top tokens to return.
            top_p: Nucleus sampling threshold.
            user: Unique user identifier.
            **kwargs: Additional provider-specific arguments.

        Yields:
            ChatCompletionChunk objects.

        """
        # This yield is required for abstract async generators
        if False:
            yield ChatCompletionChunk(
                id="",
                choices=[],
                created=0,
                model="",
                object="chat.completion.chunk",
            )

    def completion_stream(
        self,
        model: str,
        messages: Iterable[ChatCompletionMessageParam],
        *,
        frequency_penalty: float | None = None,
        logit_bias: dict[str, int] | None = None,
        logprobs: bool | None = None,
        max_completion_tokens: int | None = None,
        max_tokens: int | None = None,
        n: int | None = None,
        parallel_tool_calls: bool | None = None,
        presence_penalty: float | None = None,
        response_format: dict[str, Any] | None = None,
        seed: int | None = None,
        stop: str | list[str] | None = None,
        stream_options: dict[str, Any] | None = None,
        temperature: float | None = None,
        tool_choice: str | dict[str, Any] | None = None,
        tools: Iterable[ChatCompletionToolParam] | None = None,
        top_logprobs: int | None = None,
        top_p: float | None = None,
        user: str | None = None,
        **kwargs: Any,
    ) -> Iterator[ChatCompletionChunk]:
        """Perform a streaming chat completion request synchronously.

        Args:
            model: Model ID to use for the completion.
            messages: List of messages comprising the conversation.
            frequency_penalty: Penalize repeated tokens (-2.0 to 2.0).
            logit_bias: Map of token IDs to bias values.
            logprobs: Whether to return log probabilities.
            max_completion_tokens: Maximum tokens to generate.
            max_tokens: Legacy parameter for maximum tokens.
            n: Number of completions to generate.
            parallel_tool_calls: Enable parallel function calling.
            presence_penalty: Penalize new topics (-2.0 to 2.0).
            response_format: Format specification for structured outputs.
            seed: Seed for deterministic sampling.
            stop: Stop sequences.
            stream_options: Options for streaming.
            temperature: Sampling temperature (0 to 2).
            tool_choice: Controls tool calling behavior.
            tools: List of tools the model may call.
            top_logprobs: Number of top tokens to return.
            top_p: Nucleus sampling threshold.
            user: Unique user identifier.
            **kwargs: Additional provider-specific arguments.

        Yields:
            ChatCompletionChunk objects.

        """
        async_iter = self.acompletion_stream(
            model=model,
            messages=messages,
            frequency_penalty=frequency_penalty,
            logit_bias=logit_bias,
            logprobs=logprobs,
            max_completion_tokens=max_completion_tokens,
            max_tokens=max_tokens,
            n=n,
            parallel_tool_calls=parallel_tool_calls,
            presence_penalty=presence_penalty,
            response_format=response_format,
            seed=seed,
            stop=stop,
            stream_options=stream_options,
            temperature=temperature,
            tool_choice=tool_choice,
            tools=tools,
            top_logprobs=top_logprobs,
            top_p=top_p,
            user=user,
            **kwargs,
        )
        return async_iter_to_sync_iter(async_iter)

    @staticmethod
    @abstractmethod
    def _convert_completion_params(params: CompletionParams, **kwargs: Any) -> dict[str, Any]:
        """Convert completion parameters to provider-specific format.

        Args:
            params: The completion parameters.
            **kwargs: Additional arguments.

        Returns:
            Provider-specific parameters dictionary.

        """

    @staticmethod
    @abstractmethod
    def _convert_completion_response(response: Any) -> ChatCompletion:
        """Convert provider response to ChatCompletion format.

        Args:
            response: The provider-specific response.

        Returns:
            ChatCompletion object.

        """

    @staticmethod
    @abstractmethod
    def _convert_completion_chunk_response(response: Any, **kwargs: Any) -> ChatCompletionChunk:
        """Convert provider streaming response chunk to ChatCompletionChunk format.

        Args:
            response: The provider-specific chunk response.
            **kwargs: Additional arguments.

        Returns:
            ChatCompletionChunk object.

        """
